### ISSUE TITLE
BeOS specific Updates

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7630]
Migrated from [rt.perl.org#32717](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=32717) (status was 'resolved')

Searchable as RT32717$

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de

Hello\,

I realized that the BeOS port has become out of sync\, and I've invested a 
bit of time to deal with a couple of issues\. Attached are patches against 
the 5\.8\.5 and 5\.9\.1 sources \(identical save for file offsets\)\.

These are the changes&#8203;:

beos/beos\.c
beos/beosish\.h
\* There is no flock\(\) on BeOS and fcntl\(F\_SETLK\) doesn't work either\.
&nbsp;  I implemented a simple server that allows to provide a flock\(\)
&nbsp;  emulation\. It's behaves a bit different than flock\(\)\, but should be
&nbsp;  close enough to be useful\. close\(\) is overridden to release locks\.
\* kill\(\) doesn't behave properly when called for a process group
&nbsp;  \(i\.e\. for 0 or a value \< \-1\) and signal 0\. It really kills the
&nbsp;  processes\. This is a work\-around for this case that should work well
&nbsp;  enough\.

beos/beos\_flock\_server\.cpp
beos/beos\_flock\_server\.h
\* The flock\(\) server\.

hints/beos\.sh
\* correct identification of systems with the BONE network stack
\* added \`\-lnet' to \`libs' on non\-BONE systems&#8203;: it contains things like
&nbsp;  uname\(\)\, gethostname\(\) and is not added automatically by newer
&nbsp;  versions of gcc available for BeOS
\* override d\_flock and d\_flockproto\, since we implement an emulation

perl\.c
\* sysconf\(\_SC\_CLK\_TCK\) returns a wrong value on BeOS

ext/Errno/Errno\_pm\.PL
\* \<errno\.h> constants were ignored

ext/File/Glob/t/basic\.t
\* BeOS can't expand \`~\<username>'

lib/ExtUtils/t/MM\_BeOS\.t
\* fixed number of tests and initialized a variable to some useful value

lib/Tie/File/t/16\_handle\.t
\* $^O is \`beos' not \`BeOS'

t/op/magic\.t
\* execvp\(\) on BeOS prepends \`\./' to relative path\, which is harmless\, but
&nbsp;  let the tests fail


Remaining issues


---
1\) ext/IO/t/IO\.t

ext/IO/t/IO\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.Can't locate Socket\.pm in @&#8203;INC \(@&#8203;INC 
contains&#8203;: \.\./lib\) at \.\./lib/IO/Socket\.pm line 12\.
BEGIN failed\-\-compilation aborted at \.\./lib/IO/Socket\.pm line 12\.
Compilation failed in require at \(eval 4\) line 5\.
 at \.\./ext/IO/t/IO\.t line 38
Use of uninitialized value in unlink at \.\./ext/IO/t/IO\.t line 120\.
Use of uninitialized value in rmdir at \.\./ext/IO/t/IO\.t line 121\.
\# Looks like you planned 18 tests but only ran 4\.
\# Looks like your test died just after 4\.

The test is skipped in 5\.8\.5\, but is no longer\, due to this change between 
the versions&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- perl-5.8.5/ext/IO/t/IO.t	Wed Jun 23 17:39:25 2004
+++ perl-5.9.1/ext/IO/t/IO.t	Fri Feb 20 16:25:54 2004
@@ -4,11 +4,6 @@
 {
 	chdir 't' if -d 't';
 	@INC = '../lib';
-	require Config;
-	if ($Config::Config{'extensions'} !~ /\bSocket\b/) {
-		print "1..0 # Skip: Socket not built - IO.pm uses Socket";
-		exit 0;
-	}
 }
 
 use strict;


```
</details>
2) t/io/fflush.t
&nbsp;   ext/POSIX/t/sigaction
&nbsp;   ext/POSIX/t/waitpid\.t
\* Fail due to non\-fixable BeOS bugs\. I'm not sure what the right
&nbsp;  approach in such a situation is\. Let the tests fail or skip them?

3\) The flock\(\) server is not integrated into the build system yet\. I'm not 
familiar with the build system and thought I rather ask for assistence\. It 
can simply be compiled via \`c\+\+ \-o beos\_flock\_server beos\_flock\_server\.cpp' 
in the \`beos' subdir and would be installed in the standard bin dir\.

4\) I haven't updated the README\.beos yet\, since I'd rather wait for your 
reply first\. I'm also not quite sure how to do this best\. Apparently 
someone wrote it once and someone else added an \`Update 2002\-05\-30' 
section\. So I could rewrite the whole thing to just reflect the current 
state or I could just add another \`Update\.\.\.' section\.

Thanks in advance for your help\.

CU\, Ingo

PS&#8203;: The bug reporting email address given in

&nbsp;  http&#8203;://www\.cpan\.org/misc/cpan\-faq\.html\#How\_report\_bug\_Perl

is incorrect&#8203;: perlbug@&#8203;perl\.com instead of perlbug@&#8203;perl\.org \(two 
occurrences\)\.

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de

[perl-5.8.5.diff.bz2](https://rt-archive.perl.org/perl5/Ticket/Attachment/101307/70166/perl-5.8.5.diff.bz2)

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de

[perl-5.9.1.diff.bz2](https://rt-archive.perl.org/perl5/Ticket/Attachment/101307/70167/perl-5.9.1.diff.bz2)

**Comment by p5pRT** [comment]
### From @schwern

On Tue\, Nov 30\, 2004 at 03&#8203;:38&#8203;:32PM \-0000\, Ingo Weinhold wrote&#8203;:
> lib/ExtUtils/t/MM\_BeOS\.t
> \* fixed number of tests and initialized a variable to some useful value

Thanks\.  Caught this for the CPAN version\.


\-\- 
Michael G Schwern        schwern@&#8203;pobox\.com  http&#8203;://www\.pobox\.com/~schwern/
It's Ecstacy time\!

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tux

On Tue 30 Nov 2004 16&#8203;:38\, Ingo Weinhold \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Ingo Weinhold 
> \# Please include the string&#8203;:  \[perl \#32717\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=32717 >
> 
> 
> Hello\,
> 
> I realized that the BeOS port has become out of sync\, and I've invested a 
> bit of time to deal with a couple of issues\. Attached are patches against 
> the 5\.8\.5 and 5\.9\.1 sources \(identical save for file offsets\)\.
> 
> These are the changes&#8203;:

Thanks\, applied to blead \(5\.9\.2\) as change \#23584

It's up to the maint pumpking to integrate in 5\.8 tree

> beos/beos\.c
> beos/beosish\.h
> \* There is no flock\(\) on BeOS and fcntl\(F\_SETLK\) doesn't work either\.
>   I implemented a simple server that allows to provide a flock\(\)
>   emulation\. It's behaves a bit different than flock\(\)\, but should be
>   close enough to be useful\. close\(\) is overridden to release locks\.
> \* kill\(\) doesn't behave properly when called for a process group
>   \(i\.e\. for 0 or a value \< \-1\) and signal 0\. It really kills the
>   processes\. This is a work\-around for this case that should work well
>   enough\.
> 
> beos/beos\_flock\_server\.cpp
> beos/beos\_flock\_server\.h
> \* The flock\(\) server\.

These files are new\. Your diff shows the ols with
zero date \(1970\) instead of using /dev/null

> hints/beos\.sh
> \* correct identification of systems with the BONE network stack
> \* added \`\-lnet' to \`libs' on non\-BONE systems&#8203;: it contains things like
>   uname\(\)\, gethostname\(\) and is not added automatically by newer
>   versions of gcc available for BeOS
> \* override d\_flock and d\_flockproto\, since we implement an emulation
> 
> perl\.c
> \* sysconf\(\_SC\_CLK\_TCK\) returns a wrong value on BeOS
> 
> ext/Errno/Errno\_pm\.PL
> \* \<errno\.h> constants were ignored
> 
> ext/File/Glob/t/basic\.t
> \* BeOS can't expand \`~\<username>'
> 
> lib/ExtUtils/t/MM\_BeOS\.t
> \* fixed number of tests and initialized a variable to some useful value
> 
> lib/Tie/File/t/16\_handle\.t
> \* $^O is \`beos' not \`BeOS'
> 
> t/op/magic\.t
> \* execvp\(\) on BeOS prepends \`\./' to relative path\, which is harmless\, but
>   let the tests fail
> 
> 
> Remaining issues
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> 1\) ext/IO/t/IO\.t
> 
> ext/IO/t/IO\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.Can't locate Socket\.pm in @&#8203;INC \(@&#8203;INC 
> contains&#8203;: \.\./lib\) at \.\./lib/IO/Socket\.pm line 12\.
> BEGIN failed\-\-compilation aborted at \.\./lib/IO/Socket\.pm line 12\.
> Compilation failed in require at \(eval 4\) line 5\.
>  at \.\./ext/IO/t/IO\.t line 38
> Use of uninitialized value in unlink at \.\./ext/IO/t/IO\.t line 120\.
> Use of uninitialized value in rmdir at \.\./ext/IO/t/IO\.t line 121\.
> \# Looks like you planned 18 tests but only ran 4\.
> \# Looks like your test died just after 4\.
> 
> The test is skipped in 5\.8\.5\, but is no longer\, due to this change between 
> the versions&#8203;:
> 
> \-\-\- perl\-5\.8\.5/ext/IO/t/IO\.t    Wed Jun 23 17&#8203;:39&#8203;:25 2004
> \+\+\+ perl\-5\.9\.1/ext/IO/t/IO\.t    Fri Feb 20 16&#8203;:25&#8203;:54 2004
> @&#8203;@&#8203; \-4\,11 \+4\,6 @&#8203;@&#8203;
>  \{
>      chdir 't' if \-d 't';
>      @&#8203;INC = '\.\./lib';
> \-    require Config;
> \-    if \($Config&#8203;::Config\{'extensions'\} \!~ /\\bSocket\\b/\) \{
> \-        print "1\.\.0 \# Skip&#8203;: Socket not built \- IO\.pm uses Socket";
> \-        exit 0;
> \-    \}
>  \}
>  
>  use strict;
> 
> 2\) t/io/fflush\.t
>    ext/POSIX/t/sigaction
>    ext/POSIX/t/waitpid\.t
> \* Fail due to non\-fixable BeOS bugs\. I'm not sure what the right
>   approach in such a situation is\. Let the tests fail or skip them?
> 
> 3\) The flock\(\) server is not integrated into the build system yet\. I'm not 
> familiar with the build system and thought I rather ask for assistence\. It 
> can simply be compiled via \`c\+\+ \-o beos\_flock\_server beos\_flock\_server\.cpp' 
> in the \`beos' subdir and would be installed in the standard bin dir\.
> 
> 4\) I haven't updated the README\.beos yet\, since I'd rather wait for your 
> reply first\. I'm also not quite sure how to do this best\. Apparently 
> someone wrote it once and someone else added an \`Update 2002\-05\-30' 
> section\. So I could rewrite the whole thing to just reflect the current 
> state or I could just add another \`Update\.\.\.' section\.
> 
> Thanks in advance for your help\.
> 
> CU\, Ingo
> 
> PS&#8203;: The bug reporting email address given in
> 
>   http&#8203;://www\.cpan\.org/misc/cpan\-faq\.html\#How\_report\_bug\_Perl
> 
> is incorrect&#8203;: perlbug@&#8203;perl\.com instead of perlbug@&#8203;perl\.org \(two 
> occurrences\)\.

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.5\, & 5\.9\.x\, and 809 on  HP\-UX 10\.20 & 11\.00\, 11i\,
&nbsp;   AIX 4\.3\, AIX 5\.2\, SuSE 9\.1\, and Win2k\.  http&#8203;://www\.cmve\.net/~merijn/
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From @schwern

On Wed\, Dec 01\, 2004 at 03&#8203;:12&#8203;:36PM \+0100\, H\.Merijn Brand wrote&#8203;:
> > 4\) I haven't updated the README\.beos yet\, since I'd rather wait for your 
> > reply first\. I'm also not quite sure how to do this best\. Apparently 
> > someone wrote it once and someone else added an \`Update 2002\-05\-30' 
> > section\. So I could rewrite the whole thing to just reflect the current 
> > state or I could just add another \`Update\.\.\.' section\.

I'd say it depends on how useful the existing information is to your
average BeOS user\.  If the existing info is largely out of date\, and
you think you can do a better job\, junk it and start over\.

Since the 2002 update is just about what tests pass or fail\, and that's
likely to have changed\, and it was written by a non\-BeOS user\, it should
probably be junked and replaced with the current state of the test suite\.


\-\- 
Michael G Schwern        schwern@&#8203;pobox\.com  http&#8203;://www\.pobox\.com/~schwern/
If you haven't got the skill to do something important\, leave it alone\. 
If something is urgent or complex\, find a simple way to do it\. 
If something going wrong will particularly aggravate you\, 
&nbsp;  make certain you know how to do it\.
&nbsp;        \-\- How to beat Murphy's Law

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de


On 2004\-12\-01 at 15&#8203;:13&#8203;:14 \[\+0100\]\, you wrote&#8203;:
> On Tue 30 Nov 2004 16&#8203;:38\, Ingo Weinhold \(via RT\) 
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > \# New Ticket Created by  Ingo Weinhold
> > \# Please include the string&#8203;:  \[perl \#32717\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=32717 >
> > 
> > I realized that the BeOS port has become out of sync\, and I've invested a
> > bit of time to deal with a couple of issues\. Attached are patches against
> > the 5\.8\.5 and 5\.9\.1 sources \(identical save for file offsets\)\.
> > 
> > These are the changes&#8203;:
> 
> Thanks\, applied to blead \(5\.9\.2\) as change \#23584

Thanks\.

> > beos/beos\_flock\_server\.cpp
> > beos/beos\_flock\_server\.h
> > \* The flock\(\) server\.
> 
> These files are new\. Your diff shows the ols with
> zero date \(1970\) instead of using /dev/null

Please excuse my ignorance\, but how do get diff to output \`/dev/null' 
instead? I just used \`diff \-urN \.\.\.' \(version 2\.8\.1\) and can't find anything 
in this respect in the help text\.

CU\, Ingo

**Comment by p5pRT** [comment]
### From @tux

On Wed 01 Dec 2004 21&#8203;:22\, bonefish@&#8203;cs\.tu\-berlin\.de wrote&#8203;:
> 
> On 2004\-12\-01 at 15&#8203;:13&#8203;:14 \[\+0100\]\, you wrote&#8203;:
> > On Tue 30 Nov 2004 16&#8203;:38\, Ingo Weinhold \(via RT\) 
> > \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > > \# New Ticket Created by  Ingo Weinhold
> > > \# Please include the string&#8203;:  \[perl \#32717\]
> > > \# in the subject line of all future correspondence about this issue\.
> > > \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=32717 >
> > > 
> > > I realized that the BeOS port has become out of sync\, and I've invested a
> > > bit of time to deal with a couple of issues\. Attached are patches against
> > > the 5\.8\.5 and 5\.9\.1 sources \(identical save for file offsets\)\.
> > > 
> > > These are the changes&#8203;:
> > 
> > Thanks\, applied to blead \(5\.9\.2\) as change \#23584
> 
> Thanks\.
> 
> > > beos/beos\_flock\_server\.cpp
> > > beos/beos\_flock\_server\.h
> > > \* The flock\(\) server\.
> > 
> > These files are new\. Your diff shows the ols with
> > zero date \(1970\) instead of using /dev/null
> 
> Please excuse my ignorance\, but how do get diff to output \`/dev/null' 
> instead? I just used \`diff \-urN \.\.\.' \(version 2\.8\.1\) and can't find anything 
> in this respect in the help text\.

That's fine\. I've reread the docs\.

If you want to \*add\* a file\, you can diff against /dev/null or
pass date 0\, the latter happens on diff \-N

All the /new/ file patches I had to deal with were generated as
diff to /dev/null\.

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.5\, & 5\.9\.x\, and 809 on  HP\-UX 10\.20 & 11\.00\, 11i\,
&nbsp;   AIX 4\.3\, AIX 5\.2\, SuSE 9\.1\, and Win2k\.  http&#8203;://www\.cmve\.net/~merijn/
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de


On 2004\-12\-01 at 15&#8203;:24&#8203;:32 \[\+0100\]\, you wrote&#8203;:
> On Wed\, Dec 01\, 2004 at 03&#8203;:12&#8203;:36PM \+0100\, H\.Merijn Brand wrote&#8203;:
> > > 4\) I haven't updated the README\.beos yet\, since I'd rather wait for your
> > > reply first\. I'm also not quite sure how to do this best\. Apparently
> > > someone wrote it once and someone else added an \`Update 2002\-05\-30'
> > > section\. So I could rewrite the whole thing to just reflect the current
> > > state or I could just add another \`Update\.\.\.' section\.
> 
> I'd say it depends on how useful the existing information is to your
> average BeOS user\.  If the existing info is largely out of date\, and
> you think you can do a better job\, junk it and start over\.
> 
> Since the 2002 update is just about what tests pass or fail\, and that's
> likely to have changed\, and it was written by a non\-BeOS user\, it should
> probably be junked and replaced with the current state of the test suite\.

Thanks\, I've finally found the time to do that\. Attached is a patch against 
blead\. It also backs out the flock\(\) emulation code I previously introduced\. 
It doesn't make much sense to have it in Perl only\, so I made it a package 
and released it separately\, so that it can be used for other projects as 
well\. I adjusted the hinting for BeOS so that flock\(\) support is provided\, if 
the package is available\, respectively left out otherwise\.

The changes in detail&#8203;:

\* MANIFEST
&nbsp;  \- Removed beos/beos\_flock\_server\.cpp/\.h\.

\* Makefile\.SH
&nbsp;  \- Link miniperl on BeOS with CLDFLAGS\, too\.

\* README\.beos
&nbsp;  \- Rewritten to reflect current state\.

\* beos/beos\.c
&nbsp;  beos/beosish\.h
&nbsp;  \- Externalized flock\(\) support\.
&nbsp;  \- BeOS sigaction\(\) returns 0 in when trying to ignore or set a handler for
&nbsp;    uncatchable signals\. Added work\-around for this case\.

\* hints/beos\.sh
&nbsp;  \- Checks for external flock\(\) support and enables Perl flock\(\) if available\.

A final question&#8203;: when building Perl on BeOS executable and libraries need to 
be linked against a few libraries\. hints/beos\.sh sets \`libs' accordingly and 
this works well enough for miniperl\, perl\, and libperl\, but modules like 
POSIX don't link against those libs\, although this would be necessary\, too\. 
I've simply built everything with \`make LDLOADLIBS="\.\.\."'\, but perhaps there 
is a better way\. Is there?

CU\, Ingo

**Comment by p5pRT** [comment]
### From bonefish@cs.tu-berlin.de

<details><summary>perl-head.diff</summary>

```diff
diff -ur perl-head.orig/MANIFEST perl-head/MANIFEST
--- perl-head.orig/MANIFEST	2004-12-15 14:46:25.000000000 +0100
+++ perl-head/MANIFEST	2004-12-16 00:23:09.000000000 +0100
@@ -6,8 +6,6 @@
 av.h			Array value header
 beos/beos.c		BeOS port
 beos/beosish.h		BeOS port
-beos/beos_flock_server.cpp	BeOS port
-beos/beos_flock_server.h	BeOS port
 beos/nm.c		BeOS port
 bytecode.pl		Produces ext/ByteLoader/byterun.h, ext/ByteLoader/byterun.c and ext/B/Asmdata.pm
 cc_runtime.h		Macros need by runtime of compiler-generated code
diff -ur perl-head.orig/Makefile.SH perl-head/Makefile.SH
--- perl-head.orig/Makefile.SH	2004-12-01 20:22:07.000000000 +0100
+++ perl-head/Makefile.SH	2004-12-16 00:55:27.000000000 +0100
@@ -548,7 +548,7 @@
 !NO!SUBS!
 
 	case "${osname}${osvers}" in
-	aix*)
+	aix*|beos*)
 		$spitshell >>Makefile <<'!NO!SUBS!'
 miniperl: $& miniperlmain$(OBJ_EXT) $(LIBPERL) opmini$(OBJ_EXT)
 	$(CC) -o miniperl $(CLDFLAGS) \
@@ -557,7 +557,7 @@
 	$(LDLIBPTH) ./miniperl -w -Ilib -MExporter -e '<?>' || $(MAKE) minitest
 !NO!SUBS!
 		;;
-	beos*|next4*)
+	next4*)
 		$spitshell >>Makefile <<'!NO!SUBS!'
 miniperl: $& miniperlmain$(OBJ_EXT) $(LIBPERL) opmini$(OBJ_EXT)
 	$(CC) -o miniperl `echo $(obj) | sed 's/ op$(OBJ_EXT) / /'` \
diff -ur perl-head.orig/README.beos perl-head/README.beos
--- perl-head.orig/README.beos	2002-07-12 03:09:47.000000000 +0200
+++ perl-head/README.beos	2004-12-17 00:23:35.000000000 +0100
@@ -4,110 +4,106 @@
 
 =head1 NAME
 
-README.beos - Perl version 5 on BeOS
+README.beos - Perl version 5.8+ on BeOS
 
 =head1 DESCRIPTION
 
-Notes for building Perl under BeOS.
+This file contains instructions how to build Perl under BeOS and lists
+known problems.
 
-=head2 General Issues with Perl on BeOS
+=head1 BUILD AND INSTALL
 
-To compile perl under BeOS R4 x86:
+=head2 Requirements
 
-	./Configure -d
+I have built and tested Perl 5.8.6 and 5.9.1 under BeOS R5 x86 net server.
+I can't say anything with regard to PPC. Since Perl 5.8.0 had been released
+for BeOS BONE, I suspect, there is a good chance, that it still compiles on
+a BONE system. The only change I've made, that affects BONE systems is the
+recognition of whether it is a BONE system or not in C<hints/beos.sh>. Now
+network socket support should remain enabled on BONE systems. This might
+as well break the build, though.
 
-and hit ^C when it asks you if you want to make changes to config.sh;
-edit config.sh and do the following:
-change d_socket='define' to ='undef';
-remove SDBM, Errno, and Socket from dynamic_ext= and nonxs_ext=;
-add '#define bool short' to x2p/a2p.h;
+As more recent versions of autoconf require flock() support, I wrote a flock()
+emulation (flock_server) and released it on BeBits:
 
-	../Configure -S; make; make install
+	http://www.bebits.com/app/4030
 
-	cd ~/config/lib; ln -s 5.00502/BeOS-BePC/CORE/libperl.so .
+If you want to build a Perl with flock() support, you have to install this
+package first.
 
-(substitute 5.00502 with the appropriate filename)
+=head2 Configure
 
-=head2 BeOS Release-specific Notes
+With flock() support:
 
-=over 4
+	CFLAGS=-I/path/to/flock/server/headers ./configure.gnu \
+		--prefix=/boot/home/config
 
-=item R4 x86
+Replace C</path/to/flock/server/headers> with the path to the directory
+containing the C<flock.h> header.
 
-Dynamic loading finally works! Yay! This means you can compile your
-own modules into perl. However, Sockets and Errno still don't work.
-(Hopefully, sockets will at least work by R5, if not sooner.)
-
-=item R4 PPC
-
-I have not tested this. I rather severely doubt that dynamic loading
-will work. (My BeBox is in pieces right now, following a nasty disk
-crash.) You may have to disable dynamic loading to get the thing to
-compile at all. (use `./Configure` without -d, and say 'no' to 'Build
-a shared libperl.so'.)
+Without flock() support:
 
-=back
+	./configure.gnu --prefix=/boot/home/config
 
-=head2 Contact Information
+=head2 Build
 
-If you have comments, problem reports, or even patches or bugfixes (gasp!)
-please email me.
+With flock() support:
 
-28 Jan 1999
-Tom Spindler
-dogcow@isi.net
+	make LDLOADLIBS="-lnet -lflock"
 
-=head2 Update 2002-05-30
+Without flock() support:
 
-The following tests fail on 5.8.0 Perl in BeOS Personal 5.03:
+	make LDLOADLIBS="-lnet"
 
- t/op/lfs............................FAILED at test 17
- t/op/magic..........................FAILED at test 24
- ext/Fcntl/t/syslfs..................FAILED at test 17
- ext/File/Glob/t/basic...............FAILED at test 3
- ext/POSIX/t/sigaction...............FAILED at test 13
- ext/POSIX/t/waitpid.................FAILED at test 1
+C<-lnet> is needed on net server systems only and if the compiler doesn't
+add it automatically (Be's R5 gcc does, Oliver Tappe's gcc 2.95.3 does not).
 
-The reasons for the failures are as follows: 
+=head2 Install
 
-=over 4
+Install all perl files:
 
-=item *
+	make install
+
+Create a symlink for libperl:
 
-The t/op/lfs and ext/Fcntl/t/syslfs failures indicate that the
-LFS (large file support, files larger than 2 gigabytes) doesn't
-work from Perl (BeFS itself is well capable of supporting large
-files).  What fails is that trying to position the file pointer
-past 2 gigabytes doesn't work right, the position gets truncated
-to its lower 32 bits.
+	cd ~/config/lib; ln -s perl5/5.8.6/BePC-beos/CORE/libperl.so .
+
+Replace C<5.8.6> with your respective version of Perl.
+
+=head1 KNOWN PROBLEMS
+
+=over 4
 
 =item *
 
-The op/magic failures look like something funny going on with $0 and
-$^X that I can't now figure out: none of the generated pathnames are
-wrong as such, they just seem to accumulate "./" prefixes and infixes
-in ways that define logic.
+Network socket support is disabled for BeOS R5 net server. I didn't dare yet
+to try enabling it and see what problems occur.
 
 =item *
 
-The Glob/t/basic indicates a bug in the getpw*() functions:
-they do not always return the correct user db entries.
+The LFS (large file support) tests (C<t/op/lfs> and C<xt/Fcntl/t/syslfs>) are
+disabled as seeking beyond 2 GB is broken according to jhi@iki.fi who was the
+last one checking the BeOS port and updating this file before me. Haven't
+checked this myself.
 
 =item *
 
-The sigaction #13 means that signal mask doesn't get properly restored
-if sigaction returns early.
+The C<t/io/fflush> test fails at #6. As far as I can tell, this is caused by
+a bug in the BeOS pipes implementation that occurs when starting other child
+processes. In the particular test case a C<system("perl -e 0")> flushes the
+stdout pipe of another child process.
 
 =item *
 
-The waitpid failure means that after there are no more child
-processes, waitpid is supposed to start returning -1 (and set
-errno to ECHILD).  In BeOS, it doesn't seem to.
+The C<ext/POSIX/t/waitpid> test fails at #1. After all child processes are
+gone BeOS' waitpid(-1,...) returns 0 instead of -1 (as it should). No idea
+how to fix this.
 
 =back
 
-Disclaimer: I just installed BeOS Personal Edition 5.0 and the
-Developer Tools, that is the whole extent of my BeOS expertise,
-so please don't ask me for further help in BeOS Perl problems.
+=head1 CONTACT
+
+For BeOS specifics problems feel free to mail to:
+Ingo Weinhold <bonefish@cs.tu-berlin.de>
 
-jhi@iki.fi
+Last update: 2004-12-16
diff -ur perl-head.orig/beos/beos.c perl-head/beos/beos.c
--- perl-head.orig/beos/beos.c	2004-12-01 15:11:10.000000000 +0100
+++ perl-head/beos/beos.c	2004-12-16 17:07:10.000000000 +0100
@@ -1,9 +1,8 @@
 #include "beos/beosish.h"
-#include "beos/beos_flock_server.h"
 
 #undef waitpid
-#undef close
 #undef kill
+#undef sigaction
 
 #include <errno.h>
 #include <signal.h>
@@ -14,11 +13,6 @@
 
 #include <OS.h>
 
-/* We cache, for which FDs we got a lock. This will especially speed up close(),
-   since we won't have to contact the server. */
-#define FLOCK_TABLE_SIZE 256
-static int flockTable[FLOCK_TABLE_SIZE];
-
 /* In BeOS 5.0 the waitpid() seems to misbehave in that the status
  * has the upper and lower bytes swapped compared with the usual
  * POSIX/UNIX implementations.  To undo the surpise effect to the
@@ -34,162 +28,6 @@
     return got;
 }
 
-/* The flock() emulation worker function. */
-
-static status_t beos_flock(int fd, int operation) {
-    static int serverPortInitialized = 0;
-    static port_id serverPort = -1;
-
-    struct stat st;
-    int blocking;
-    port_id replyPort;
-    sem_id lockSem = -1;
-    status_t error;
-    flock_server_request request;
-    flock_server_reply *reply = NULL;
-
-    if (fd < 0)
-        return B_BAD_VALUE;
-
-    blocking = !(operation & LOCK_NB);
-    operation &= LOCK_SH | LOCK_EX | LOCK_UN;
-
-    /* don't try to unlock something that isn't locked */
-    if (operation == LOCK_UN && fd < FLOCK_TABLE_SIZE && !flockTable[fd])
-        return B_OK;
-
-    /* if not yet initialized, get the server port */
-    if (!serverPortInitialized) {
-        serverPort = find_port(FLOCK_SERVER_PORT_NAME);
-        /* bonefish: If the port wasn't present at this point, we could start
-         * the server. In fact, I tried this and in works, but unfortunately
-         * it also seems to confuse our pipes (with both load_image() and
-         * system()). So, we can't help it, the server has to be started
-         * manually. */
-        serverPortInitialized = ~0;
-    }
-    if (serverPort < 0)
-        return B_ERROR;
-
-    /* stat() the file to get the node_ref */
-    if (fstat(fd, &st) < 0)
-        return errno;
-
-    /* create a reply port */
-    replyPort = create_port(1, "flock reply port");
-    if (replyPort < 0)
-        return replyPort;
-
-    /* create a semaphore others will wait on while we own the lock */
-    if (operation != LOCK_UN) {
-        char semName[64];
-        sprintf(semName, "flock %ld:%lld\n", st.st_dev, st.st_ino);
-        lockSem = create_sem(0, semName);
-        if (lockSem < 0) {
-            delete_port(replyPort);
-            return lockSem;
-        }
-    }
-
-    /* prepare the request */
-    request.replyPort = replyPort;
-    request.lockSem = lockSem;
-    request.device = st.st_dev;
-    request.node = st.st_ino;
-    request.fd = fd;
-    request.operation = operation;
-    request.blocking = blocking;
-
-    /* We ask the server to get us the requested lock for the file.
-     * The server returns semaphores for all existing locks (or will exist
-     * before it's our turn) that prevent us from getting the lock just now.
-     * We block on them one after the other and after that officially own the
-     * lock. If we told the server that we don't want to block, it will send
-     * an error code, if that is not possible. */
-
-    /* send the request */
-    error = write_port(serverPort, 0, &request, sizeof(request));
-
-    if (error == B_OK) {
-        /* get the reply size */
-        int replySize = port_buffer_size(replyPort);
-        if (replySize < 0)
-            error = replySize;
-
-        /* allocate reply buffer */
-        if (error == B_OK) {
-            reply = (flock_server_reply*)malloc(replySize);
-            if (!reply)
-                error = B_NO_MEMORY;
-        }
-
-        /* read the reply */
-        if (error == B_OK) {
-            int32 code;
-            ssize_t bytesRead = read_port(replyPort, &code, reply, replySize);
-            if (bytesRead < 0) {
-                error = bytesRead;
-            } else if (bytesRead != replySize) {
-                error = B_ERROR;
-            }
-        }
-    }
-
-    /* get the error returned by the server */
-    if (error == B_OK)
-        error = reply->error;
-
-    /* wait for all lockers before us */
-    if (error == B_OK) {
-        int i;
-        for (i = 0; i < reply->semaphoreCount; i++)
-            while (acquire_sem(reply->semaphores[i]) == B_INTERRUPTED);
-    }
-
-    /* free the reply buffer */
-    free(reply);
-
-    /* delete the reply port */
-    delete_port(replyPort);
-
-    /* on failure delete the semaphore */
-    if (error != B_OK)
-        delete_sem(lockSem);
-
-    /* update the entry in the flock table */
-    if (error == B_OK && fd < FLOCK_TABLE_SIZE) {
-        if (operation == LOCK_UN)
-            flockTable[fd] = 0;
-        else
-            flockTable[fd] = 1;
-    }
-
-    return error;
-}
-
-/* We implement flock() using a server. It is not really compliant with, since
- * it would be very hard to track dup()ed FDs and those cloned as side-effect
- * of fork(). Our locks are bound to the process (team) and a particular FD.
- * I.e. a lock acquired by a team using a FD can only be unlocked by the same
- * team using exactly the same FD (no other one pointing to the same file, not
- * even when dup()ed from the original one). close()ing the FD releases the
- * lock (that's why we need to override close()). On termination of the team
- * all locks owned by the team will automatically be released. */
-
-int flock(int fd, int operation) {
-    status_t error = beos_flock(fd, operation);
-    return (error == B_OK ? 0 : (errno = error, -1));
-}
-
-/* We need to override close() to release a potential lock on the FD. See
-   flock() for details */
-
-int beos_close(int fd) {
-    flock(fd, LOCK_UN);
-
-    return close(fd);
-}
-
 
 /* BeOS kill() doesn't like the combination of the pseudo-signal 0 and
  * specifying a process group (i.e. pid < -1 || pid == 0). We work around
@@ -210,3 +48,20 @@
 
     return kill(pid, sig);
 }
+
+/* sigaction() should fail, if trying to ignore or install a signal handler
+ * for a signal that cannot be caught or ignored. The BeOS R5 sigaction()
+ * doesn't return an error, though. */
+int beos_sigaction(int sig, const struct sigaction *act,
+                   struct sigaction *oact)
+{
+    int result = sigaction(sig, act, oact);
+
+    if (result == 0 && act && act->sa_handler != SIG_DFL
+        && act->sa_handler != SIG_ERR && (sig == SIGKILL || sig == SIGSTOP)) {
+        result = -1;
+        errno = EINVAL;
+    }
+
+    return result;
+}
diff -ur perl-head.orig/beos/beosish.h perl-head/beos/beosish.h
--- perl-head.orig/beos/beosish.h	2004-12-01 15:11:10.000000000 +0100
+++ perl-head/beos/beosish.h	2004-12-16 16:46:03.000000000 +0100
@@ -11,24 +11,28 @@
 /* This seems to be protoless. */
 char *gcvt(double value, int num_digits, char *buffer);
 
+/* flock support, if available */
+#ifdef HAS_FLOCK
 
-/* flock() operation flags */
-#define LOCK_SH	(0x00)
-#define LOCK_EX	(0x01)
-#define LOCK_UN	(0x02)
-#define LOCK_NB	(0x04)
-
-int flock(int fd, int operation);
+#include <flock.h>
 
 #undef close
-#define close beos_close
+#define close flock_close
+
+#undef dup2
+#define dup2 flock_dup2
 
-int beos_close(int fd);
+#endif /* HAS_FLOCK */
 
 
 #undef kill
 #define kill beos_kill
 int beos_kill(pid_t pid, int sig);
 
+#undef sigaction
+#define sigaction(sig, act, oact) beos_sigaction((sig), (act), (oact))
+int beos_sigaction(int sig, const struct sigaction *act,
+                   struct sigaction *oact);
+
 #endif
 
diff -ur perl-head.orig/hints/beos.sh perl-head/hints/beos.sh
--- perl-head.orig/hints/beos.sh	2004-12-01 15:11:10.000000000 +0100
+++ perl-head/hints/beos.sh	2004-12-16 00:10:24.000000000 +0100
@@ -48,9 +48,27 @@
 	libs='-lnet'
 fi
 
-# We provide a flock() emulation.
-d_flock='define'
-d_flockproto='define'
+# There's a third party flock() emulation. Check, if it is available.
+echo "#include <flock.h>" > try.c
+if cc -E $CFLAGS try.c 2> /dev/null | grep "flock.*("; then
+    d_flock='define'
+    d_flockproto='define'
+    libs="$libs -lflock"
+    ldflags="$ldflags -L/boot/home/config/lib"
+else
+	cat << 'EOM' >&4
+
+I couldn't find a <flock.h> header defining a flock() prototype. That header
+comes with the flock server package (available on BeBits). You have to add
+the path to the directory containing the header via the environment variable
+CFLAGS (should contain -I</path/to/dir/of/flock/header>). Perl will be compiled
+without flock() support, if the flock server package is not installed or the
+header not found.
+
+EOM
+
+fi
+rm try.c
 
 ld='gcc'
 
```
</details>

**Comment by p5pRT** [comment]
### From @rgs

bonefish@&#8203;cs\.tu\-berlin\.de wrote&#8203;:
\.\.\. 
> Thanks\, I've finally found the time to do that\. Attached is a patch against 
> blead\. It also backs out the flock\(\) emulation code I previously introduced\. 
> It doesn't make much sense to have it in Perl only\, so I made it a package 
> and released it separately\, so that it can be used for other projects as 
> well\. I adjusted the hinting for BeOS so that flock\(\) support is provided\, if 
> the package is available\, respectively left out otherwise\.

Thanks\, applied as change \#23661\.

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Dec 01\, 2004 at 03&#8203;:12&#8203;:36PM \+0100\, H\.Merijn Brand wrote&#8203;:

> Thanks\, applied to blead \(5\.9\.2\) as change \#23584
> 
> It's up to the maint pumpking to integrate in 5\.8 tree

I've now integrated everything from blead to maint\, so barring unexpected
calamities it will be in 5\.8\.7

Nicholas Clark

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
